File: techbuzz-backend/python-services/common/publisher.py
import json
from datetime import datetime

def publish_event(redis_client, stream_name, event_type, aggregate_id, producer, payload):
    """
    Publishes a business event to a Redis Stream with a standard envelope.
    """
    event_id = f"{event_type}:{aggregate_id}"
    
    envelope = {
        "eventId": event_id,
        "eventType": event_type,
        "eventVersion": "1.0",
        "occurredAt": datetime.utcnow().isoformat() + "Z",
        "aggregateId": aggregate_id,
        "producer": producer,
        "data": json.dumps(payload)
    }

    # xadd with MAXLEN ~ 500k to prevent unbounded memory growth
    redis_client.xadd(
        stream_name,
        fields=envelope,
        maxlen=500000,
        approximate=True
    )
    
    return event_id

File: techbuzz-backend/backend/src/events/publisher.js




const { getRedisClient } = require('../config/redis');


async function publishEvent(streamName, eventType, aggregateId, producer, payload) {
  const redis = await getRedisClient();
  const eventId = `${eventType}:${aggregateId}`;
  
  const envelope = {
    eventId,
    eventType,
    eventVersion: '1.0',
    occurredAt: new Date().toISOString(),
    aggregateId,
    producer,
    data: JSON.stringify(payload)
  };

  
  await redis.xAdd(
    streamName,
    '*', 
    {
      eventId: envelope.eventId,
      eventType: envelope.eventType,
      eventVersion: envelope.eventVersion,
      occurredAt: envelope.occurredAt,
      aggregateId: envelope.aggregateId,
      producer: envelope.producer,
      data: envelope.data
    },
    {
      TRIM: {
        strategy: 'MAXLEN',
        strategyModifier: '~',
        threshold: 500000
      }
    }
  );

  return eventId;
}

module.exports = { publishEvent };
